README.md
# ff-node-monitor [![Build Status](https://travis-ci.com/freifunk-saar/ff-node-monitor.svg?branch=master)](https://travis-ci.com/freifunk-saar/ff-node-monitor)

This is a simple web service for Freifunk networks that lets node operators
register to monitor their nodes.  It uses the `nodes.json` from
[hopglass](https://github.com/hopglass/hopglass) to detect which nodes are
online, and sends notifications when the online status changes.

## Setup

The setup consists of three parts: Getting the service built, getting the
database set up, and configuring the service to be run as a daemon.

I have tested the following steps on a Debian Stretch system; if you are using a
different version or a different distribution, you might have to change some of
the steps accordingly.

Make sure you have at least 1.5 GB free disk space.

### Build Process

1.  First, let's create a user for this service, and change to its home directory:

    ```
    sudo adduser ff-node-monitor --home /opt/ff-node-monitor --system
    ```

2.  We need some development libraries for the build process:

    ```
    sudo apt install libssl-dev libpq-dev libc6-dev curl gcc pkg-config
    ```

3.  Fetch the *ff-node-monitor* sources:

    ```
    sudo -u ff-node-monitor git clone https://github.com/freifunk-saar/ff-node-monitor.git ~ff-node-monitor/src
    cd ~ff-node-monitor/src
    ```

4.  *ff-node-monitor* is written in [Rust](https://www.rust-lang.org/) using
    [Rocket](https://rocket.rs/). Any recent stable version of Rust should work:

    ```
    curl https://sh.rustup.rs -sSf > rustup.sh
    sudo -u ff-node-monitor sh rustup.sh
    rm rustup.sh
    ```

5.  Build *ff-node-monitor*:

    ```
    sudo -u ff-node-monitor ~ff-node-monitor/.cargo/bin/cargo build --release
    ```

    > The build process takes a while, you can already finish the Database Setup
    > (steps 7. and 8.) and part of the Service Setup (step 9. and 11.) in a
    > second shell as the build process continues.

6.  This step is optional, but if you want to save some disk space, you can now
    clean up the build directory:

    ```
    rm -rf target/release/{build,deps,incremental,.fingerprint}
    ```

    Over time, you will also accumulate more and more different Rust versions.
    You can use

    ```
    sudo -u ff-node-monitor ~ff-node-monitor/.cargo/bin/rustup toolchain list
    ```

    to see which versions you have installed, and then `toolchain uninstall`
    the ones you do not need any more (all but the last, most likely).

### Database setup

7.  *ff-node-monitor* needs PostgreSQL as a database backend:

    ```
    sudo apt install postgresql
    ```

8.  We will use the `ff-node-monitor` system user to access PostgreSQL, and we
    need to create a database for the service:

    ```
    sudo -u postgres psql -c 'CREATE ROLE "ff-node-monitor" WITH LOGIN;'
    sudo -u postgres psql -c 'CREATE DATABASE "ff-node-monitor" WITH OWNER = "ff-node-monitor" LC_COLLATE = '\''de_DE.utf8'\'' TEMPLATE template0;'
    ```

    You may have to install the `de_DE.UTF-8` locale before this works.  On
    Debian, run `sudo dpkg-reconfigure locales` to do so.

### Service setup

9.  The service loads its configuration from a `Rocket.toml` file in the source
    directory.  You can start by copying the template:

    ```
    cd ~ff-node-monitor/src
    sudo -u ff-node-monitor cp Rocket.toml.dist Rocket.toml
    chmod 600 Rocket.toml
    ```

    Most of the values in your `Rocket.toml` will need to be changed; see the comments in the
    template for what to do and how.

10. To run the service using systemd, the `.service` file needs to be installed:

    ```
    sudo cp ff-node-monitor.service /etc/systemd/system/
    sudo systemctl daemon-reload
    sudo systemctl enable ff-node-monitor
    sudo systemctl start ff-node-monitor
    sudo systemctl status ff-node-monitor
    ```

    If the last command does not show the service as running, you need to debug
    and fix whatever issue has come up.

11. To expose the service on the internet, set up a reverse proxy in your main
    web server. If you are not already running a web server, `nginx` is a good
    choice.  You will have to edit your site configuration, usually located at
    `/etc/nginx/sites-enabled/default`.  Here's the necessary snippet, mounting
    the node monitor in the `node-monitor` subdirectory:

    ```
    location /node-monitor/ {
        proxy_pass http://127.0.0.1:8833/;
    }
    # Directly serve static files, no need to run them through the app
    location /node-monitor/static/ {
        alias /opt/ff-node-monitor/src/static/;
    }
    ```

    Test your configuration and reload nginx:

    ```
    nginx -t
    service nginx reload
    ```

    Now, accessing the service at whatever `root` URL you configured in the
    `Rocket.toml` should work.

12. Finally, the service relies on a cron job to regularly check in on all the
    nodes and send notifications when their status changed:

    ```
    sudo crontab -e -u ff-node-monitor
    ```

    Add the following line to that crontab, replacing `$ROOT_URL` by your `root` URL
    (as configured in `Rocket.toml`):

    ```
    */5 * * * *    curl -s $ROOT_URL/cron
    ```

That's it!  The service should now be running and working.

## Upgrade

To upgrade the service to the latest git version, follow these steps:

```
cd ~ff-node-monitor/src/
git pull
sudo rm target/release/ff-node-monitor
sudo -u ff-node-monitor ~ff-node-monitor/.cargo/bin/cargo build --release
sudo systemctl restart ff-node-monitor
```

Check [the CHANGELOG](CHANGELOG.md) to see if any manual steps are needed.

## Debugging

When something goes wrong, the first step should be to look at the error log:

```
sudo journalctl -u ff-node-monitor.service
```

## Customization

If you want to adapt the node monitor to the layout of your web presence, you
can set `stylesheet` to an external CSS file in your `Rocket.toml`.
Put any images and other static data into `src/static/`. Here a CSS-example:

```
#title {
    padding-top: 131px;
    background: url('static/logo.png');
    background-repeat: no-repeat;
}
```

## Development Virtual Environment

You can easily set up a test VM using Vagrant.

If you want to tweak the default configuration (which you do not have to), first
copy the default config file `vagrant.config.dist` to `vagrant.config` and
customize the configuration. Then follow these steps:

```
sudo apt install vagrant
cd vagrant
vagrant up ff-service
```

This will take a while, downloading the Vagrant box and install a running system
inside. You can adapt `bootstrap.sh` as you like to test around with different
settings. In your real setup you at least have to change the root URL where you
will be hosting ff-node-monitor.

You can then access the vagrant box at `http://localhost:8833`. If you want to
login the server use

```
vagrant ssh ff-service
```

If you want to delete and start over use

```
vagrant destroy ff-service
vagrant up ff-service
```

If you want to send out emails, one easy option is [msmtp](https://wiki.debian.org/msmtp).

For gmail (with deactivated 2-factor login) use this configuration in `/etc/msmtprc`:

```
# Set default values for all following accounts.
defaults
port 587
tls on
tls_trust_file /etc/ssl/certs/ca-certificates.crt

account gmail
host smtp.gmail.com
from <user>@gmail.com
auth on
user <user>
password <your password>

# Set a default account
account default : gmail
```

To test it run

```
echo -e "Subject: msmtp test\nhello test." | msmtp _recipient_address_
```

You should find your sent e-mail in the recipient's inbox shortly afterwards.

### vagrant troubleshooting

In case the install in the mashine fails, consider to upgrade your vagrant box with

```
vagrant box update
```


CHANGELOG.md
## 2023-12-31

* We updated to Rocket v0.5. This is almost entirely an internal change, but it has two user-visible consequences:
  The `ROCKET_ENV` environment variable no longer has any effect. You should remove it from your systemd service file.
  (There is a new `ROCKET_PROFILE` environment variable with a similar effect, but it should not be needed
  for usual deployments. Just make sure you build in release mode, i.e., `cargo build --release` or `cargo install`).
  Furthermore, you can now use stable Rust to build ff-node-monitor; a nightly version is no longer required.

## 2018-12-10

* We switched to Rocket's built-in support for managing DB connection pools.  You have to **update your `Rocket.toml`**:
  Remove `postgres_url = "postgres://ff-node-monitor@/ff-node-monitor"` from the `[global.ff-node-monitor.secrets]` section, and add `postgres = { url = "postgres://ff-node-monitor@/ff-node-monitor" }` to the `[global.databases]` section.


vagrant/bootstrap.sh
#!/usr/bin/env bash
# bootstrap script to install ff-node-monitor
set -xe

# in case vagrant fails to install, try run this script starting from line 25

export DEBIAN_FRONTEND=noninteractive

if [ "$0" == "/tmp/vagrant-shell" ]; then
  cd /vagrant
else
  cd vagrant
fi

: "#### include default config or user config, that is in .gitignore:"
if [ -f vagrant.config ]; then
  source vagrant.config
else
  source vagrant.config.dist
fi

# some variables used in this script

# port and URL
PORT=8833 # matches the port forwarding in Vagrantfile
ROOT_URL="http://localhost:$PORT"

HOME_PATH='/opt/ff-node-monitor'

FFNM_USERNAME="ff-node-monitor"

: "#### create a user for this service, and change to its home directory:"
sudo adduser $FFNM_USERNAME --home "$HOME_PATH" --system
cd "$HOME_PATH"
sudo chown $FFNM_USERNAME .
ffsudo="sudo -u $FFNM_USERNAME"

: "#### We need some development libraries for the build process:"
sudo apt update && sudo apt -y upgrade && sudo apt -y --no-install-recommends install git msmtp libc6-dev libssl-dev libpq-dev curl gcc pkg-config

: "#### setup locale"
sed -i 's/# \(\(de_DE\|en_US\)\.UTF-8 UTF-8\)/\1/' /etc/locale.gen && dpkg-reconfigure --frontend=noninteractive locales

: "#### fetch the ff-node-monitor sources:"
# do not fail when re-provisioning
test -d src || $ffsudo git clone https://github.com/freifunk-saar/ff-node-monitor.git src

: "#### ff-node-monitor is written in Rust using Rocket, which means it needs a nightly version of Rust:"
if ! test -f $HOME_PATH/.cargo/bin/rustc; then
    curl https://sh.rustup.rs -sSf | $ffsudo sh -s -- -y
fi
cd "$HOME_PATH/src"
$ffsudo "$HOME_PATH/.cargo/bin/cargo" build --release

: "#### Database setup"
apt -y install postgresql
psql="sudo -u postgres psql"
if ! $psql -lqt | cut -d \| -f 1 | grep -qw ff-node-monitor; then
    $psql -c 'DROP ROLE IF EXISTS "'$FFNM_USERNAME'"; CREATE ROLE "'$FFNM_USERNAME'" WITH LOGIN;'
    $psql -c 'CREATE DATABASE "'$FFNM_USERNAME'" WITH OWNER = "'$FFNM_USERNAME'" LC_COLLATE = '\''de_DE.utf8'\'' TEMPLATE template0;'
fi

: "#### Service setup"
: "#### The service loads its configuration from a Rocket.toml file in the source directory. You can start by copying the template"
cd "$HOME_PATH/src"
$ffsudo touch Rocket.toml
$ffsudo chmod 777 Rocket.toml
cat <<EOF > Rocket.toml
[global.ff-node-monitor.ui]
# The name and grammatical article of this instance.
instance_name = "$INSTANCE_NAME"
# The sentence "Willkommen bei $instance_article_dative $instance_name" should be grammatically
# correct.
instance_article_dative = "der"
# The sender address of the emails that are sent by ff-node-monitor.
email_from = "$EMAIL_FROM"
# Optional: When fewer nodes than this are online in the nodes.json file, the database won't be updated
# and there will be no warning mails sent. If set, this should be at least as high as number of
# gateways in the network (to handle the case where only those are shown as online).
#min_online_nodes = 10

[global.ff-node-monitor.urls]
# The root URL where you will be hosting ff-node-monitor (with trailing slash)
root = "$ROOT_URL"
# The URL to the hopglass nodes.json file.
nodes = "$NODES_URL"
# URL to the source code (needed for AGPL compliance).  You can leave this unchanged if you didn't
# change the code.  Otherwise, you have to upload the changed code somewhere and point to it here.
sources = "https://github.com/freifunk-saar/ff-node-monitor"
# Optional: Absolute URL to another stylesheet that is included in the page.
#stylesheet = "https://..."

[global.ff-node-monitor.secrets]
# Key used to sign data for confirmation emails.  Generate this key with
# `openssl rand -hex 32`.
action_signing_key = "$(openssl rand -hex 32)"
# Optional: Host to submit emails to.  That host must accept email with arbitrary destination
# from this service.  Unless this is "localhost", the connection will be encrypted via STARTTLS.
smtp_host = "localhost"

[global.databases]
postgres = { url = "postgres://$FFNM_USERNAME@/$FFNM_USERNAME" }

[global]
# The address on which ff-node-monitor will listen.  Use "0.0.0.0" to listen on all interfaces.
address = "0.0.0.0"
port = $PORT
# Secret key used by Rocket.  Generate this key with `openssl rand -base64 32`.
secret_key = "$(openssl rand -base64 32)"
EOF
$ffsudo chmod 600 Rocket.toml

: "#### To run the service using systemd, the .service file needs to be installed:"
sudo cp ff-node-monitor.service /etc/systemd/system/
: "#### If you change the HOME_PATH, you also have to adapt the path in the service file"
sudo sed -i 's|/opt/ff-node-monitor|'"$HOME_PATH"'|g' /etc/systemd/system/ff-node-monitor.service
sudo systemctl daemon-reload
sudo systemctl stop ff-node-monitor
sudo systemctl enable ff-node-monitor.service
sudo systemctl start ff-node-monitor
sudo systemctl status ff-node-monitor

: "#### Finally, the service relies on a cron job to regularly check in on all the nodes and send notifications when their status changed:"
(sudo crontab -u $FFNM_USERNAME -l; echo "*/5 * * * *    curl -s $ROOT_URL/cron" ) | sudo crontab -u $FFNM_USERNAME -

: "#### read node data initially:"
sleep 1
$ffsudo curl -sS $ROOT_URL/cron

echo "The site should now be reacheable under $ROOT_URL"
